scipy/special/xsf/tools.h

```diff
@@ -261,10 +258,167 @@ namespace detail {
      * See ContinuedFractionSeriesGenerator for details.
      */
     template <typename Generator, typename T = pair_value_t<generator_result_t<Generator>>>
-    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T>
-    continued_fraction_series(Generator &cf) {
+    XSF_HOST_DEVICE ContinuedFractionSeriesGenerator<Generator, T> continued_fraction_series(Generator &cf) {
         return ContinuedFractionSeriesGenerator<Generator, T>(cf);
     }
 
+    /* Find initial bracket for a bracketing scalar root finder. A valid bracket is a pair of points a < b for
+     * which the signs of f(a) and f(b) differ. If f(x0) = 0, where x0 is the initial guess, this bracket finder
+     * will return the bracket (x0, x0). It is expected that the rootfinder will check if the bracket
+     * endpoints are roots.
+     *
+     * This is a private function intended specifically for the situation where
+     * the goal is to invert a CDF function F for a parametrized family of distributions with respect to one
+     * parameter, when the other parameters are known, and where F is monotonic with respect to the unknown parameter.
+     *
+     * Note that this takes a pointer to a function taking a tuple of args along with a scalar
+     * double argument. A tuple of args for specializing func is also passed as the final argument.
+     * It would be much cleaner to use std::function and capturing lambda's to specialize the
+     * function we are finding a root for, but I couldn't get this working in CuPy using NVRTC.
+     * This should be revisited in the future in order to allow simplifying this code. */
+    template <typename... Args>
+    XSF_HOST_DEVICE inline std::tuple<double, double, double, double, int> bracket_root_for_cdf_inversion(
+        double (*func)(double, std::tuple<Args...>), double x0, double xmin, double xmax, double step0_left,
```

Comment:
So we definitely shouldn't use a function pointer here. There are two ways to do generic callbacks in C++. One if `std::function`, which is mostly meant for dynamic situations. That involves virtual functions, we don't want that.

The other is simply adding a template parameter for the callable itself. Here you could do that with `template <typename Function, ...>` and the argument then being `bracket_root_for_cdf_inversion((Function func, ...)`. Then it works for a callable, be it a lambda, function pointer, or whatever. I did this exactly in `legendre.h`, but it's also done in the standard library in the `<algorithms>` header.